### PR TITLE
Modified GC so deleted data can be dropped #559

### PR DESCRIPTION
**Pull request by keith-turner**
The Oracle is now the autority on GC.  The Oracle looks in Zookeeper to
determine the oldest active timestamp for all transactors.   After finding the
oldest active timestamp if persist this in zookeeper for use by garbage
collection iterators.  The Oracle also returns the GC timestamp it persisted to
Zookeeper to any clients that request a timestamp.  Clients use this GC time
from the oracle when committing or closing a snapshot to detemine if they
possibly read stale data.

### CODE REVIEW

File: modules/accumulo/src/main/java/io/fluo/accumulo/util/ZookeeperUtil.java [hunk 1]
```diff
@@ -59,54 +59,19 @@ public static String parseRoot(String zookeepers) {
...
-   * 
+   *
    * @param zookeepers Zookeeper connection string
    * @return Oldest active timestamp or oldest possible ts (-1) if not found
    */
-  public static long getOldestTimestamp(String zookeepers) {
-    long oldestTs = Long.MAX_VALUE;
-    boolean nodeFound = false;
-
+  public static long getGcTimestamp(String zookeepers) {
```
Comment (mikewalch) [inline modules/accumulo/src/main/java/io/fluo/accumulo/util/ZookeeperUtil.java]: Can you update or remove the docs for the method?

File: modules/accumulo/src/main/java/io/fluo/accumulo/iterators/GarbageCollectionIterator.java [hunk 2]
```diff
@@ -87,7 +90,7 @@ public void init(SortedKeyValueIterator<Key, Value> source, Map<String, String>
       if (zookeepers == null) {
         throw new IllegalArgumentException("A configuration item for GC iterator was not set");
       }
-      oldestActiveTs = ZookeeperUtil.getOldestTimestamp(zookeepers);
+      oldestActiveTs = ZookeeperUtil.getGcTimestamp(zookeepers);
```
Comment (mikewalch) [inline modules/accumulo/src/main/java/io/fluo/accumulo/iterators/GarbageCollectionIterator.java]: Would changing this variable name to gcTimestamp make this code more clear?  I guess the gcTimestamp is technically the oldest active timestamp but it's a little confusing.

File: modules/accumulo/src/main/java/io/fluo/accumulo/util/ZookeeperUtil.java [hunk 3]
```diff
@@ -59,54 +59,19 @@ public static String parseRoot(String zookeepers) {
...
-      }
-
-      // Return oldest possible timestamp if no node found
-      if (!nodeFound) {
-        return OLDEST_POSSIBLE;
-      }
-      return oldestTs;
-    } catch (IOException e) {
-      throw new IllegalStateException(e);
+      byte[] d = zk.getData(ZookeeperPath.ORACLE_GC_TIMESTAMP, false, null);
```
Comment (mikewalch) [inline modules/accumulo/src/main/java/io/fluo/accumulo/util/ZookeeperUtil.java]: I like how this code is really simple now.

File: modules/accumulo/src/main/java/io/fluo/accumulo/values/WriteValue.java [hunk 4]
```diff
@@ -17,16 +17,16 @@
 import io.fluo.accumulo.util.ByteArrayUtil;
 
 /**
- * 
+ *
```
Comment (mikewalch) [inline modules/accumulo/src/main/java/io/fluo/accumulo/values/WriteValue.java]: Can you add a comment or remove the empty block?

File: modules/core/src/main/java/io/fluo/core/oracle/Stamp.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  private final long stamp;
+  private final long gcStamp;
+
+  Stamp(long stamp, long gcStamp) {
+    this.stamp = stamp;
+    this.gcStamp = gcStamp;
+  }
+
+
+  public long getTimestamp() {
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/oracle/Stamp.java]: you could rename this to `getTxTimestamp()` to differentiate between `getGcTimestamp()`

Comment (keith-turner) [inline modules/core/src/main/java/io/fluo/core/oracle/Stamp.java]: sounds good, I did that in 4d0872b

File: modules/accumulo/src/main/java/io/fluo/accumulo/values/WriteValue.java (hunk 4 above)
Comment (keith-turner) [inline modules/accumulo/src/main/java/io/fluo/accumulo/values/WriteValue.java]: done in 4d0872b

File: modules/accumulo/src/main/java/io/fluo/accumulo/iterators/GarbageCollectionIterator.java (hunk 2 above)
Comment (keith-turner) [inline modules/accumulo/src/main/java/io/fluo/accumulo/iterators/GarbageCollectionIterator.java]: done in 4d0872b

File: modules/accumulo/src/main/java/io/fluo/accumulo/util/ZookeeperUtil.java (hunk 1 above)
Comment (keith-turner) [inline modules/accumulo/src/main/java/io/fluo/accumulo/util/ZookeeperUtil.java]: done in 4d0872b

File: modules/core/src/main/java/io/fluo/core/oracle/Stamp.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ * 
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.fluo.core.oracle;
+
+public class Stamp {
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/oracle/Stamp.java]: For consistency, could rename all variables from `stamp` to `txStamp`

File: modules/core/src/main/thrift/oracle.thrift [hunk 7]
```diff
@@ -11,10 +11,14 @@
  * or implied. See the License for the specific language governing permissions and limitations under
  * the License.
  */
-namespace java io.fluo.impl.thrift
+namespace java io.fluo.core.thrift
 
+struct Stamps {
+  1: i64 stampStart,
```
Comment (mikewalch) [inline modules/core/src/main/thrift/oracle.thrift]: could name this `txStamp`

File: modules/core/src/main/java/io/fluo/core/oracle/OracleClient.java [hunk 8]
```diff
@@ -173,7 +176,9 @@ private void doWork() {
 
               final Context timerContext = responseTimer.time();
 
-              start = localClient.getTimestamps(env.getFluoApplicationID(), request.size());
+              Stamps stamps = localClient.getTimestamps(env.getFluoApplicationID(), request.size());
+              start = stamps.stampStart;
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/oracle/OracleClient.java]: could name this `txStamp`

Comment (keith-turner) [inline modules/core/src/main/java/io/fluo/core/oracle/OracleClient.java]: done in a87ce16

File: modules/core/src/main/thrift/oracle.thrift (hunk 7 above)
Comment (keith-turner) [inline modules/core/src/main/thrift/oracle.thrift]: done in a87ce16

File: modules/core/src/main/java/io/fluo/core/oracle/Stamp.java (hunk 6 above)
Comment (keith-turner) [inline modules/core/src/main/java/io/fluo/core/oracle/Stamp.java]: done in a87ce16

**Comment by mikewalch** [comment]
+1

**Pull request merged**